File: external/cnem_03-10-17/matlab/tuto3d/tuto_interpol3d.m
clc;
clear all;
close all;

addpath('../bin');

%% load domain (nodes + boundary(s)) and creat dummy  then plot

load('data_interpol_1');

%% creat dummy point location for interpolation evaluation

XYZ_Point=rand(1000,3)*1.5-0.75;

%% plot

figure;
hold on;
tri_out_handle=trimesh(IN_Tri_Ini,XYZ_Noeud(:,1),XYZ_Noeud(:,2),XYZ_Noeud(:,3),'edgecolor','black');
alpha(tri_out_handle,0.5);
plot3(XYZ_Noeud(:,1),XYZ_Noeud(:,2),XYZ_Noeud(:,3),'.','color','green');
plot3(XYZ_Point(:,1),XYZ_Point(:,2),XYZ_Point(:,3),'o','color','blue');
axis vis3d
axis equal
hold off;

%% initialize the interpolator

Type_FF = 0;
%IN_Tri_Ini
Interpol=m_cnem3d_interpol(XYZ_Noeud,[],XYZ_Point,Type_FF);

%% interpolate a filds
% test : Var = XYZ_Noeud ==> interpolated Var on XYZ_Point = XYZ_Point

Var=XYZ_Noeud;
%Var_Int=Interpol.interpolate(Var);
Var_Int=Interpol.mat_interpol_glob*Var;

%% cal error

dif=Var_Int-XYZ_Point;
j=0;
ind_p_in=zeros(sum(Interpol.In_Out),1);
for i=1:size(Interpol.In_Out,1)
    if Interpol.In_Out(i)
        j=j+1;
        ind_p_in(j)=i;
    end
end

int_p_out=setdiff(1:size(Interpol.In_Out,1),ind_p_in);

err=max(max(abs(dif(ind_p_in,:))))

%% plot in out point

figure;
hold on;
tri_out_handle=trimesh(IN_Tri_Ini,XYZ_Noeud(:,1),XYZ_Noeud(:,2),XYZ_Noeud(:,3),'edgecolor','black');
alpha(tri_out_handle,0.5);
plot3(XYZ_Point(ind_p_in,1),XYZ_Point(ind_p_in,2),XYZ_Point(ind_p_in,3),'*','color','red');
plot3(XYZ_Point(int_p_out,1),XYZ_Point(int_p_out,2),XYZ_Point(int_p_out,3),'*','color','blue');
axis vis3d
axis equal
hold off;

File: external/cnem_03-10-17/python/tuto/tuto_interpol3d.py
##
## Copyright (C) 2003-2011
## Lounes ILLOUL (illoul_lounes@yahoo.fr)
## Philippe LORONG (philippe.lorong@ensam.eu)
## Arts et Metiers ParisTech, Paris, France
##
## $Revision: 25 $ - $Date: 2012-02-22 17:36:30 +0100 (mer., 22 fevr. 2012) $

import shutil
import re
import math
import cPickle
import numpy as np
import scipy as sp

import sys
sys.path.append('../bin')
import load_node_front
import py_cnem3d

#-------------------------------------------------------------------------------
# read nodes and boundary :
print '\n--------------------------------------------------------------------------------\n'
print 'read node and boundary...\n'

[XYZ_Noeud,IN_Tri_Ini]=load_node_front.load3d('cubeuni3d/')

#-------------------------------------------------------------------------------
# creat dummy point location for interpolation evaluation

XYZ_Point=list((np.random.random(1000000*3)-0.5)*0.15)

#-------------------------------------------------------------------------------
# creat the interpolator
print '\n--------------------------------------------------------------------------------\n'
print 'creat the interpolator...\n'


Type_FF=0

interpol=py_cnem3d.interpol(XYZ_Noeud,IN_Tri_Ini,XYZ_Point,Type_FF)

## if we don't enter the boundary tri IN_Tri_Ini, the convex hull of the nodes
##    will be taken as boundary, let's try :

#interpol=py_cnem3d.interpol(XYZ_Noeud,[],XYZ_Point,Type_FF)

#----------------------------------------------------------------------------
# interpolate variable filds : Var
print '\n--------------------------------------------------------------------------------\n'
print 'interpolate variable filds...\n'

# test : Var = XYZ_Noeud ==> interpolated Var on XYZ_Point = XYZ_Point

Var=np.array(XYZ_Noeud)
Var=Var.reshape((Var.shape[0]/3,3))

#interpolate Var :
#Var_Int=interpol.interpolate(Var)

Mat_Int=interpol.mat_interpol_glob()
Var_Int=Mat_Int*Var

XYZ_Point=np.array(XYZ_Point)
XYZ_Point=XYZ_Point.reshape((XYZ_Point.shape[0]/3,3))

Dif=Var_Int-XYZ_Point

maxdif=0.

for i in xrange(len(Dif)):
    if interpol.In_Out[i] : # point interpolation in the domain 
        maxdif_i=max(abs(Dif[i]))
        if maxdif_i>maxdif:
            maxdif=maxdif_i
        
print 'err max : ', maxdif

#-------------------------------------------------------------------------------
